### PR TITLE
Archive labels UI

### PR DESCRIPTION
**Pull request by puni9869**
Followup  https://github.com/go-gitea/gitea/pull/26478
## Archived labels UI

Added:

* Implemented the structural framework for the user interface to accommodate archived labels.
* Introduced archived labels functionality to the UI, akin to the approach taken for closed milestones or projects. This functionality allows concealing information while retaining it, without resorting to deletion.
* Particularly advantageous for managing outdated labels that are no longer in use, offering a means to preserve them without cluttering the active label set.

Changed:
* Enhanced the Filtered UI page to seamlessly incorporate a list of archived labels.
* Introduced sorting capability specifically for archived labels.

Outsourced:
* Defer the implementation of specialized handling for archived labels to upcoming pull requests. This step will be undertaken subsequent to the successful merge of this pull request.

Note:
The Label page now showcases archived labels distinctly from unarchived/active labels. Sorting functionality is also extended to both categories, enhancing user experience and management of labels.

### When no labels
image

### When all labels are unarchived/active
image

### When we have archived and unarchived labels both but unarchived/active label tab is active
image

### When we have archived and unarchived labels both but unarchived label tab is active
image

### Marking again archived label unarchive.
image

### Sort Archived Label `reversealphabetically`
image

image

image

### Tool Tip added for new feature indication near Archived Label Count on both archived and active/unarchived page
image

---

Part of https://github.com/go-gitea/gitea/issues/25237

### CODE REVIEW

File: models/issues/label.go [hunk 1]
```diff
@@ -113,10 +113,10 @@ func (l *Label) CalOpenIssues() {
 
 // SetArchived set the label as archived
 func (l *Label) SetArchived(isArchived bool) {
-	if isArchived && l.ArchivedUnix.IsZero() {
-		l.ArchivedUnix = timeutil.TimeStampNow()
-	} else {
```
Comment (delvh) [inline models/issues/label.go:118]: Yep, you're absolutely right.
I missed that edge case.

File: models/issues/label.go [hunk 2]
```diff
@@ -113,10 +113,10 @@ func (l *Label) CalOpenIssues() {
...
+	if !isArchived {
 		l.ArchivedUnix = timeutil.TimeStamp(0)
+	} else if isArchived && l.ArchivedUnix.IsZero() {
```
Comment (delvh) [inline models/issues/label.go:118]:
```suggestion
	} else if l.ArchivedUnix.IsZero() { // Only change the date when it is newly archived.
```

**Comment by puni9869** [comment]
This PR was rebased to https://github.com/go-gitea/gitea/pull/26478. So thats why it contains previous commits. 
We can squash merge.

File: templates/repo/issue/labels/label_list.tmpl [hunk 3]
```diff
@@ -9,10 +32,11 @@
...
-					<a class="{{if eq .SortType "mostissues"}}active {{end}}item" href="{{$.Link}}?sort=mostissues&state={{$.State}}">{{.locale.Tr "repo.milestones.filter_sort.most_issues"}}</a>
+
+					<a class="{{if or (eq .SortType "alphabetically") (not .SortType)}}active {{end}}item" href="{{$.Link}}?sort=alphabetically&state={{$.State}}&archived_label={{if .IsPageArchivedLabels}}true{{end}}">{{.locale.Tr "repo.issues.label.filter_sort.alphabetically"}}</a>
+					<a class="{{if eq .SortType "reversealphabetically"}}active {{end}}item" href="{{$.Link}}?sort=reversealphabetically&state={{$.State}}&archived_label={{if .IsPageArchivedLabels}}true{{end}}">{{.locale.Tr "repo.issues.label.filter_sort.reverse_alphabetically"}}</a>
+					<a class="{{if eq .SortType "leastissues"}}active {{end}}item" href="{{$.Link}}?sort=leastissues&state={{$.State}}&archived_label={{if .IsPageArchivedLabels}}true{{end}}">{{.locale.Tr "repo.milestones.filter_sort.least_issues"}}</a>
+					<a class="{{if eq .SortType "mostissues"}}active {{end}}item" href="{{$.Link}}?sort=mostissues&state={{$.State}}&archived_label={{if .IsPageArchivedLabels}}true{{end}}">{{.locale.Tr "repo.milestones.filter_sort.most_issues"}}</a>
```
Comment (delvh) [inline templates/repo/issue/labels/label_list.tmpl:39]:
```suggestion
					<a class="{{if or (eq .SortType "alphabetically") (not .SortType)}}active {{end}}item" href="{{$.Link}}?sort=alphabetically&state={{$.State}}{{if .IsPageArchivedLabels}}&archived_label=true{{end}}">{{.locale.Tr "repo.issues.label.filter_sort.alphabetically"}}</a>
					<a class="{{if eq .SortType "reversealphabetically"}}active {{end}}item" href="{{$.Link}}?sort=reversealphabetically&state={{$.State}}{{if .IsPageArchivedLabels}}&archived_label=true{{end}}">{{.locale.Tr "repo.issues.label.filter_sort.reverse_alphabetically"}}</a>
					<a class="{{if eq .SortType "leastissues"}}active {{end}}item" href="{{$.Link}}?sort=leastissues&state={{$.State}}{{if .IsPageArchivedLabels}}&archived_label=true{{end}}">{{.locale.Tr "repo.milestones.filter_sort.least_issues"}}</a>
					<a class="{{if eq .SortType "mostissues"}}active {{end}}item" href="{{$.Link}}?sort=mostissues&state={{$.State}}{{if .IsPageArchivedLabels}}&archived_label=true{{end}}">{{.locale.Tr "repo.milestones.filter_sort.most_issues"}}</a>
```

File: templates/repo/issue/labels/label_list.tmpl
Comment (delvh) [inline templates/repo/issue/labels/label_list.tmpl:1]: Hmm… I've thought about what the best design for it is.
I think the winner is the following design:
grafik
(That can be created with the following changes:
```html
<h4 class="ui top attached header gt-df">
  <span class="gt-f1">18 labels</span>
  <label class="gt-f1"><input type="checkbox" class="show-archived-labels-checkbox"> Show archived labels</label>
  <div class="gt-f1">…
</h4>
```
)
Where we only show the middle checkbox if there are archived labels.
It needs the following JS pseudocode:
```js
document.querySelector('.show-archived-labels-checkbox').addEventListener('input',
  e => for(const archivedLabel of document.querySelectorAll('.archived-label'))
    toggleElem(archivedLabel, e.target.checked)
  )
);
```
We should highlight the archived labels, i.e. by using a different background for them.
Other than that, archived labels should not be treated differently on this page.

Comment (delvh) [reply templates/repo/issue/labels/label_list.tmpl:1]: That way, we don't need a separate page, hide unnecessary information by default, add no technical debt but still allow configuring anything the user wants to configure.

Comment (puni9869) [reply templates/repo/issue/labels/label_list.tmpl:1]: Correct, this UI is good and easy to use. here is the implementation https://github.com/go-gitea/gitea/pull/26511/files

**Comment by puni9869** [comment]
Fix moved in https://github.com/go-gitea/gitea/pull/26511/files

**Pull request closed without merging**